src/net/db.rs

```diff
@@ -0,0 +1,1029 @@
+        let conn = Connection::open_in_memory()
+            .map_err(|e| db_error::SqliteError(e))?;
+
+        let mut db = PeerDB {
+            conn: conn,
+            readwrite: true,
+        };
+
+        db.instantiate(network_id, key_expires, asn4_entries, initial_neighbors)?;
+        Ok(db)
+    }
+
+    pub fn conn(&self) -> &Connection {
+        &self.conn
+    }
+    
+    pub fn tx_begin<'a>(&'a mut self) -> Result<Transaction<'a>, db_error> {
+        if !self.readwrite {
+            return Err(db_error::ReadOnly);
+        }
+
+        let tx = self.conn.transaction()
+            .map_err(|e| db_error::SqliteError(e))?;
+        Ok(tx)
+    }
+
+    /// Read the local peer record 
+    pub fn get_local_peer(conn: &DBConn) -> Result<LocalPeer, db_error> {
+        let row_order = LocalPeer::row_order().join(",");
+        let qry = format!("SELECT {} FROM local_peer LIMIT 1", row_order);
+        let rows = query_rows::<LocalPeer, _>(conn, &qry.to_string(), NO_PARAMS)?;
+
+        match rows.len() {
+            1 => Ok(rows[0].clone()),
+            _ => {
+                // only one item here 
+                panic!("Got multiple LocalPeer rows, or 0");
+            }
+        }
+    }
+
+    /// Set the local IP address and port 
+    pub fn set_local_ipaddr<'a>(tx: &mut Transaction<'a>, addrbytes: &PeerAddress, port: u16) -> Result<(), db_error> {
+        tx.execute("UPDATE local_peer SET addrbytes = ?1, port = ?2", &[&to_hex(&addrbytes.as_bytes().to_vec()), &port as &ToSql])
+            .map_err(db_error::SqliteError)?;
+
+        Ok(())
+    }
+
+    /// Set local service availability 
+    pub fn set_local_services<'a>(tx: &mut Transaction<'a>, services: u16) -> Result<(), db_error> {
+        tx.execute("UPDATE local_peer SET services = ?1", &[&services as &ToSql])
+            .map_err(db_error::SqliteError)?;
+
+        Ok(())
+    }
+
+    /// Set local private key and expiry 
+    pub fn set_local_private_key<'a>(tx: &mut Transaction<'a>, privkey: &Secp256k1PrivateKey, expire_block: u64) -> Result<(), db_error> {
+        if expire_block > ((1 as u64) << 63) - 1 {
+            return Err(db_error::Overflow);
+        }
+
+        tx.execute("UPDATE local_peer SET private_key = ?1, private_key_expire = ?2",
+                   &[&to_hex(&privkey.to_bytes()), &(expire_block as i64) as &ToSql])
+            .map_err(db_error::SqliteError)?;
+
+        Ok(())
+    }
+
+    /// Calculate the "slots" in the peer database where this peer can be inserted.
+    /// Slots are distributed uniformly at random between 0 and 2**24.
+    /// NUM_SLOTS will be returned.
+    pub fn peer_slots(conn: &DBConn, network_id: u32, peer_addr: &PeerAddress, peer_port: u16) -> Result<Vec<u32>, db_error> {
+        let local_peer = PeerDB::get_local_peer(conn)?;
+        let mut ret = vec![];
+        for i in 0..NUM_SLOTS {
```

Comment:
this loop would probably be clearer (and more efficient) if it just built 1 mutable byte array, and then modified the 33rd byte and hashed the array

Reply:
(i.e., you hoist the rest of the byte array construction out of the loop)